homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,625 @@
+"""Support for interfacing with the HASS MPRIS agent."""
+
+from __future__ import annotations
+
+import asyncio
+from collections.abc import AsyncGenerator
+import json
+import re
+from typing import Any, Final, Union, cast
+
+from hassmpris.proto import mpris_pb2
+import hassmpris_client
+
+from homeassistant.components.media_player import (
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+    MediaPlayerState,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    ATTR_PLAYBACK_RATE,
+    DOMAIN,
+    EXPECTED_HEARTBEAT_FREQUENCY,
+    LOGGER as _LOGGER,
+)
+
+PLAYER_STATE_MAP: dict[Any, MediaPlayerState] = {
+    mpris_pb2.PlayerStatus.GONE: MediaPlayerState.OFF,  # type: ignore[attr-defined]
+    mpris_pb2.PlayerStatus.APPEARED: MediaPlayerState.IDLE,  # type: ignore[attr-defined]
+    mpris_pb2.PlayerStatus.PLAYING: MediaPlayerState.PLAYING,  # type: ignore[attr-defined]
+    mpris_pb2.PlayerStatus.PAUSED: MediaPlayerState.PAUSED,  # type: ignore[attr-defined]
+    mpris_pb2.PlayerStatus.STOPPED: MediaPlayerState.IDLE,  # type: ignore[attr-defined]
+}
+UNAVAILABLE = "unavailable"
+
+SUPPORTED_MINIMAL = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.STOP
+)
+SUPPORTED_TURN_OFF = MediaPlayerEntityFeature.TURN_OFF
+SUPPORTED_TURN_ON = MediaPlayerEntityFeature.TURN_ON
+
+# Enable to remove clones (second / third, nth instances)
+# of any player when the player exits or the integration
+# reconnects to an agent and these instances of the player
+# are no longer running.
+# Not enabled by default because events in the logbook
+# disappear when the player is removed.
+# This will become a setting configurable through the UI
+# once it's figured out how to add settings.
+REMOVE_CLONES_WHILE_RUNNING = False
```

Comment:
There is no bug in the integration.

It is normal — though definitely a corner case — that multiple entities, representing one player each, are created, when multiple players with the same name are running.  The default, as you can see with the variable, is to keep an entity per player opened.  "Clones" are simply multiple instances of a player with the same name.  Some players support running multiple instances — VLC is the stereotypical example.

The idea behind this feature, when it is enabled, is to remove player entities when instances [n, n-1..., 2] of these players exit, only leaving one player entity of each kind behind (managed by the instance of the integration) in the Home Assistant entity registry.  If the boolean is on, and then you start five VLC players, and then you close the last 4 VLCs opened, only one VLC player remains (in state off).  Of course, if you open Amarok, Clementine, Totem, and VLC, and you exit all of them, entities remain for each one of them behind, in state `OFF`.

This is intended to be a setting.  I just haven't gotten around to actually providing an UI point for this setting — thinking that perhaps it would be easier to get this code reviewed if the code could be limited to the absolute minimum.

I will remove the code but leave a comment for myself to add that setting later.  It's a two-line deletion — not sure what we gain with it — but I'm happy to comply.

Reply:
I have added an options flow with tests.

Reply:
Now that this code is justified by the existence of the setting (and the setting is justified by the user experience) I can mark this as resolved.

Reply:
OK, thank you for the explanation 👍